### DISCUSSION TITLE
Rumble

### DISCUSSION THREAD

**Discussion by Socke81** [Etaash-mathamsetty/trigger-control#13] [Ideas]
You can delete it. It's probably my code which is why Rumble doesn't work. 
Original message:
_Hi, your app works perfectly. Can you also implement rumble? SDL2 always tells me in my app that haptic is not available for the dualsense and therefore I can't use the rumble functions. I briefly looked into your code and saw that you completely ignore this and just send some bytes to the gamepad. Could you also do that to use the rumble function or does that not work at all on the PC?_

**Comment by BananaWorks07** [comment]
It generally works fine on Linux, if you are on windows it won't work without enabling the HIDAPI SDL flag

**Reply by Socke81** [reply]
I hope I'm not annoying you. But do you know how it looks like with this special Dualsense Rumble on PC? Is that possible? It happened to me once that Windows recognized my Dualsense as an audio device and played the currently running radio stream.

**Reply by BananaWorks07** [reply]
I'm not sure what you are asking but you can go read the source code for hid-playstation.c in the linux kernel, it contains the inner workings for how the controller works. 
The dualsense's ability to play rumble based on audio stream is a feature intentionally created by sony, all you have to do is play audio that is designed to rumble the controller in a certain way. (this won't work through bluetooth)
On bluetooth the most you get is to just use SDL's built in rumble features (the controller will emulate DS4 rumble)

**Reply by Socke81** [reply]
Yes, I mean this "audio rumble". I just connected the Dualsense via cable and a "wireless controller" audio device popped up. So unfortunately only possible by cable. Don't know if it makes sense to implent this. Thanks for your help.

**Reply by BananaWorks07** [reply]
> Yes, I mean this "audio rumble". I just connected the Dualsense via cable and a "wireless controller" audio device popped up. So unfortunately only possible by cable. Don't know if it makes sense to implent this. Thanks for your help.

This is easy to do using SDL, just set your SDL audio output as the controller and play whatever sound you like using SDL. 
I don't think it makes sense to add it to this project, since libtrigger-control only supports triggers and nothing else (everything else in the GUI uses APIs built into SDL)

**Reply by Socke81** [reply]
Is the Unreal Engine. :) But there you can also select the audio device. I will try it out later. The Windows test sound makes the pad vibrate. But with 1-2 seconds delay. Have so far implemented Rumble and LED. Am just about to implent the touchpad and the sensors then I have to debug your code to implent the triggers.

**Reply by BananaWorks07** [reply]
> Is the Unreal Engine. :) But there you can also select the audio device. I will try it out later. The Windows test sound makes the pad vibrate. But with 1-2 seconds delay. Have so far implemented Rumble and LED. Am just about to implent the touchpad and the sensors then I have to debug your code to implent the triggers.

That's cool, I was thinking about doing this with unreal engine, then I remembered I kinda hate it (worst linux support ever imaginable) and decided I would write my own engine one day. Since, Unity is really slow (on linux) and Godot doesn't have C++ support. (C# is good enough tbh, so I'll revist)

**Reply by Socke81** [reply]
I have implented the triggers into my plugin. But I used other bytes than you. According to this page you should not do it the way you did it.
https://gist.github.com/Nielk1/6d54cc2c00d2201ccb8c2720ad7538db

My code now works on Windows 10 and Mac. But not on Ubuntu. But your code works on Linux or? Haven't tried your app on Linux. Which SDL version did you use? And does the mouse move when you use the tochpad? I also don't get sensor (gyro) values in SDL on Ubuntu. I can't change LED color on Ubuntu either.

**Reply by BananaWorks07** [reply]
you need ubuntu 22.04 for this to work, I added a note about this in the readme a long time ago
also this is the way you do it according to DS4 windows, I just followed what they did and what pydualsense did, it's also what the linux kernel does, (this code is based on an SDL2 demo that the SDL2 devs have on their git) I allowed the user to control every single aspect of the triggers that's somewhat safe to use and is currently supported. Obviously there might be foot guns here and there

**Reply by Socke81** [reply]
I got the link from the SDL source code. I think it does not feel different on the end. That's why the official effects are enough for me. I don't want to implement anything that could be removed by a firmware update. Makes no difference to your app anyway I think. 

I use SDL 2.0.20 on Linux because Epic uses it in Unreal Engine 5.0. On Windows and Mac I can use my own SDL version. On Linux unfortunately not. 

I think the Unreal Engine is not that bad on Linux. After all there is even a compiled version from Epic. You just have to unpack it.

**Reply by BananaWorks07** [reply]
> I got the link from the SDL source code. I think it does not feel different on the end. That's why the official effects are enough for me. I don't want to implement anything that could be removed by a firmware update. Makes no difference to your app anyway I think.
> 
> I use SDL 2.0.20 on Linux because Epic uses it in Unreal Engine 5.0. On Windows and Mac I can use my own SDL version. On Linux unfortunately not.
> 
> I think the Unreal Engine is not that bad on Linux. After all there is even a compiled version from Epic. You just have to unpack it.

no it's pretty bad, c++ is unusable, and right click menus dont work properly in blueprints, it's a complete piece of garbage
anyway, as to why your code isn't working on Linux, my answer is idk, it works fine for me on my pc, it also should work fine on sdl 2.0.20, since it is plenty new enough

**Reply by Socke81** [reply]
Oh, that doesn't sound good. I've only tested my plugins briefly so far, so I haven't been able to try much. Blueprints worked normally though. I tried Visual Studio code briefly and that worked fine. I will test my game project tomorrow just for fun. 

Ah, forgot to write. I use the drivers from NVIDA.

**Reply by BananaWorks07** [reply]
> Oh, that doesn't sound good. I've only tested my plugins briefly so far, so I haven't been able to try much. Blueprints worked normally though. I tried Visual Studio code briefly and that worked fine. I will test my game project tomorrow just for fun.
> 
> Ah, forgot to write. I use the drivers from NVIDA.

might work for you since you are using ubuntu, but I seriously wish people stop testing things against ubuntu, it's got pretty outdated packages even on the newest release and it leads to stuff like the glibc fiasco happening. They should also support at least 2 distros only supporting ubuntu (ubuntu based counts) is like supporting 30% of the linux community and if they support arch (arch based also counts) as well, that's an additional 30%!
